tool/tsh/resolve_default_addr.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+type raceResult struct {
+	addr string
+	err  error
+}
+
+// logResponseBody reads and dumps a response body to the log at the supplied
+// level. Note that it is still the caller's responsibility to close the body
+// stream.
+func logResponseBody(level logrus.Level, bodyStream io.Reader) {
+	if log.Logger.Level < level {
+		return
+	}
+
+	// NB: `ReadAll()` will time out (or be cancelled) according to the
+	//     context originally supplied to the request that initiated this
+	//     response, so no need to have an independent reading timeout
+	//     here.
+	body, err := ioutil.ReadAll(bodyStream)
+	if err != nil {
+		// This is only for debugging purposes, so it's safe to just give up here.
+		log.WithError(err).Debug("Failed to read body stream")
+		return
+	}
+
+	log.Logf(level, "Response body: %q", body)
+}
+
+// raceRequest drives an HTTP request to completion and posts the results back
+// to the supplied channel.
+func raceRequest(ctx context.Context, cli *http.Client, addr string, waitgroup *sync.WaitGroup, results chan<- raceResult) {
+	defer waitgroup.Done()
+
+	target := fmt.Sprintf("https://%s/webapi/ping", addr)
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, target, nil)
+
+	if err == nil {
```

Comment:
I would invert all `err == nil` checks which would make the code easier to reason about and reduce indentation, even if with a little redundancy.

```go
func raceRequest(ctx context.Context, cli *http.Client, addr string, waitgroup *sync.WaitGroup, results chan<- raceResult) {
  defer waitgroup.Done()

  target := fmt.Sprintf("https://%s/webapi/ping", addr)
  request, err := http.NewRequestWithContext(ctx, http.MethodGet, target, nil)
  if err != nil {
    results <- raceResult{addr: addr, err: err}
    return
  }

  rsp, err := cli.Do(request)
  if err != nil {
    results <- raceResult{addr: addr, err: err}
    return
  }
  defer rsp.Body.Close()
  
  if rsp.StatusCode != http.StatusOK {
    // ...
    return
  }

  results <- raceResult{addr: addr}
}
```

Reply:
Done